### PR TITLE
kvm: Use lscpu to get cpu max speed

### PR DESCRIPTION
**Pull request by davidjumani**
### Description

Use lscpu to get the cpu max frequency
Fixes https://github.com/apache/cloudstack/issues/5497

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [x] Minor

### CODE REVIEW

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 1]
```diff
@@ -85,8 +87,18 @@ protected static long getCpuSpeed(final NodeInfo nodeInfo) {
             LOGGER.info(String.format("Retrieved value [%s] from file [%s]. This corresponds to a CPU speed of [%s] MHz.", cpuInfoMaxFreq, cpuInfoMaxFreqFileName, cpuInfoMaxFreq / 1000));
             return cpuInfoMaxFreq / 1000;
         } catch (IOException | NumberFormatException e) {
-            LOGGER.error(String.format("Unable to retrieve the CPU speed from file [%s]. Using the value [%s] provided by the Libvirt.", cpuInfoMaxFreqFileName, nodeInfo.mhz), e);
-            return nodeInfo.mhz;
+            try {
+                LOGGER.error(String.format("Unable to retrieve the CPU speed from file [%s]. Trying to fetch it from lscpu",
+                    cpuInfoMaxFreqFileName), e);
+                String command = "lscpu | grep -i 'Model name' | head -n 1 | egrep -o '[[:digit:]].[[:digit:]]+GHz' | sed 's/GHz//g'";
```
Comment (rhtyd) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:93]: @davidjumani I think standard way is to get from /proc/cpuinfo; `grep "^[c]pu MHz" /proc/cpuinfo`

Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:93]: @rhtyd @davidjumani 

there are multiple ways to get the cpu speed. 

while(true);do cat /proc/cpuinfo |grep 'cpu MHz' |head -n1;sleep 0.1;done
while(true);do virsh nodeinfo |grep 'CPU frequency';sleep 0.1;done

it looks the result by `lscpu` is most stable. I am not sure if it works for all cpu models.

it seems we should not get cpu speed from `/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq`. If cpu supports Intel Turbo Boost Technology, it is the max cpu frequency not the normal cpu frequency.  for example, on my laptop
```
$ cat /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq
4700000
$ virsh nodeinfo |grep 'CPU frequency'
CPU frequency:       2800 MHz
$ cat /proc/cpuinfo |grep 'cpu MHz' |head -n1
cpu MHz		: 2800.000
```

my opinion is: lscpu > nodeInfo.mhz (libvirt-java) -> /proc/cpuinfo > cpuinfo_max_freq

Comment (davidjumani) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:93]: @weizhouapache I'll rearrange the CPU speed fetching stages

**Comment by davidjumani** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
tested ok on nested u20 environement.

@leolleeooleo could you please test it ?

**Comment by weizhouapache** [comment]
@blueorangutan test

**Comment by leolleeooleo** [comment]
> tested ok on nested u20 environement.
> 
> @leolleeooleo could you please test it ?

lscpu works fine on Intel CPU.
So that on the Intel platform can get normal frequency from lscpu.

AMD CPU doesn't have frequency in "Model name".
So that on AMD platform will get the maximum frequency from /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq.
I don't have AMD server to try it, but that will get the normal frequency from cpuinfo_max_freq I had search from google.

Now all CPUs have boost frequency (Intel® Turbo Boost Technology and AMD Turbo Core).
How do you think the server CPU capacity is depend on normal frequency or maximum frequency?

If you think Turbo frequency can get more performance on VMs,
I suggest change the order to:
cpuinfo_max_freq > lscpu > libvirt

**Comment by weizhouapache** [comment]
> > tested ok on nested u20 environement.
> > @leolleeooleo could you please test it ?
> 
> lscpu works fine on Intel CPU. So that on the Intel platform can get normal frequency from lscpu.
> 
> AMD CPU doesn't have frequency in "Model name". So that on AMD platform will get the maximum frequency from /sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq. I don't have AMD server to try it, but that will get the normal frequency from cpuinfo_max_freq I had search from google.
> 
> Now all CPUs have boost frequency (Intel® Turbo Boost Technology and AMD Turbo Core). How do you think the server CPU capacity is depend on normal frequency or maximum frequency?
> 
> If you think Turbo frequency can get more performance on VMs, I suggest change the order to: cpuinfo_max_freq > lscpu > libvirt

@leolleeooleo 
lscpu, libvirt, /proc/cpuinfo return almost same value, but cpuinfo_max_freq returns different value if cpu supports Turbo boost or turbo core. 
as far as I know, only single core can reach the max frequency. we cannot assume all cpu cores can reach the max frequency.

**Comment by leolleeooleo** [comment]
@weizhouapache 
Got it, you're right.
Only single core can reach the max frequency.

**Comment by weizhouapache** [comment]
> @weizhouapache Got it, you're right. Only single core can reach the max frequency.

@davidjumani @rhtyd @leolleeooleo @nvazquez 
can you get a consensus what's the best order ?

my opinion is: lscpu (might not work on dell servers) > nodeInfo.mhz (libvirt-java) = /proc/cpuinfo > cpuinfo_max_freq

**Comment by leolleeooleo** [comment]
@weizhouapache 

lscpu work on dell servers
libvirt don't work on dell servers.

Dell PowerEdge R740 :
```
[leo@kvm1 ~]$ lscpu 
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              32
On-line CPU(s) list: 0-31
Thread(s) per core:  2
Core(s) per socket:  8
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) Silver 4208 CPU @ 2.10GHz                     <<<<< here is ok
Stepping:            7
CPU MHz:             1634.801                                                         <<<<< low frequency
CPU max MHz:         3200.0000
CPU min MHz:         800.0000
BogoMIPS:            4200.00
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            11264K
NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities

[leo@kvm1 ~]$ sudo virsh nodeinfo 
CPU model:           x86_64
CPU(s):              32
CPU frequency:       1006 MHz                                            <<<<< low frequency
CPU socket(s):       1
Core(s) per socket:  8
Thread(s) per core:  2
NUMA cell(s):        2
Memory size:         65364380 KiB


[leo@kvm1 ~]$ cat /proc/cpuinfo
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 85
model name      : Intel(R) Xeon(R) Silver 4208 CPU @ 2.10GHz                         <<<<< here is ok
stepping        : 7
microcode       : 0x5003102
cpu MHz         : 823.050                                                           <<<<< low frequency
cache size      : 11264 KB
physical id     : 0
siblings        : 16
core id         : 0
cpu cores       : 8
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 22
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
bugs            : spectre_v1 spectre_v2 spec_store_bypass swapgs taa itlb_multihit
bogomips        : 4200.00
clflush size    : 64
cache_alignment : 64
address sizes   : 46 bits physical, 48 bits virtual
power management:

```

**Comment by davidjumani** [comment]
@blueorangutan help

**Comment by davidjumani** [comment]
Agree with the order @weizhouapache proposed

**Comment by s-seitz** [comment]
As @weizhouapache mentioned earlier ("lscpu > nodeInfo.mhz (libvirt-java) -> /proc/cpuinfo > cpuinfo_max_freq"), I'ld suggest to outsource the whole logic into a script (at a tbd .../scripts/ .../scriptlets/-folder) to ensure ENV's are set correctly. Keep in mind "lscpu" is usually localized, so at least PATH and LC_ALL are necessary. In my opinion, having a script with defined return-values in place, would also make testing and debugging a lot easier.

**Comment by sureshanaparti** [comment]
ping @davidjumani Is this PR ready?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
> ping @davidjumani Is this PR ready?

hi @davidjumani any update on this PR?

**Comment by davidjumani** [comment]
@sureshanaparti Apologies for the delay. I've addressed most of the comments

**Comment by davidjumani** [comment]
cpuInfoMaxFreqFileName

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan test

**Review by sureshanaparti — APPROVED** [review]
code LGTM

**Review by sureshanaparti — APPROVED** [review]

**Comment by sureshanaparti** [comment]
@blueorangutan test

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 2]
```diff
@@ -80,16 +82,39 @@ public long getOverCommitMemory() {
...
+        try {
+            return getCpuSpeedByLSCPU();
+        } catch (NullPointerException | NumberFormatException e) {
+            LOGGER.error(String.format("Unable to retrieve the CPU speed from lscpu. Using the value in [%s].", cpuInfoMaxFreqFileName), e);
+            return getCpuSpeedFromFile();
+        }
+    }
+
+    private static long getCpuSpeedByLSCPU() {
+        LOGGER.info("Fetching cpu speed from it from lscpu");
```
Comment (GutoVeronezi) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:103]:
```suggestion
        LOGGER.info("Fetching CPU speed from command \"lscpu\".");
```

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 3]
```diff
@@ -80,16 +82,39 @@ public long getOverCommitMemory() {
...
+    private static long getCpuSpeedInternal() throws IOException {
+        try {
+            return getCpuSpeedByLSCPU();
+        } catch (NullPointerException | NumberFormatException e) {
+            LOGGER.error(String.format("Unable to retrieve the CPU speed from lscpu. Using the value in [%s].", cpuInfoMaxFreqFileName), e);
+            return getCpuSpeedFromFile();
+        }
+    }
+
+    private static long getCpuSpeedByLSCPU() {
```
Comment (GutoVeronezi) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:102]:
```suggestion
    private static long getCpuSpeedFromCommandLscpu() {
```

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 4]
```diff
@@ -80,16 +82,39 @@ public long getOverCommitMemory() {
...
+        LOGGER.info("Fetching cpu speed from it from lscpu");
+        String command = "lscpu | grep -i 'Model name' | head -n 1 | egrep -o '[[:digit:]].[[:digit:]]+GHz' | sed 's/GHz//g'";
+        String result = Script.runSimpleBashScript(command);
+        long speed = (long) (Float.parseFloat(result) * 1000);
+        LOGGER.info(String.format("Got [%d] speed from lscpu", speed));
```
Comment (GutoVeronezi) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:107]:
```suggestion
        LOGGER.info(String.format("Command [%s] resulted in the value [%s] for CPU speed.", command, speed));
```

**Comment by sureshanaparti** [comment]
@davidjumani can you address the outstanding review comments.

**Comment by davidjumani** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 5]
```diff
@@ -80,16 +82,39 @@ public long getOverCommitMemory() {
...
+            return getCpuSpeedInternal();
+        } catch (IOException | NumberFormatException ex) {
+            LOGGER.error(String.format("Unable to retrieve the CPU speed from file [%s] and lscpu. Using the value [%s] provided by the Libvirt.", cpuInfoMaxFreqFileName, nodeInfo.mhz), ex);
             return nodeInfo.mhz;
         }
     }
 
+    private static long getCpuSpeedInternal() throws IOException {
+        try {
+            return getCpuSpeedByLSCPU();
```
Comment (sureshanaparti) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:95]:
```suggestion
            return getCpuSpeedFromCommandLscpu();
```

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 6]
```diff
@@ -80,13 +82,23 @@ public long getOverCommitMemory() {
...
 
     protected static long getCpuSpeed(final NodeInfo nodeInfo) {
-        try (Reader reader = new FileReader(cpuInfoMaxFreqFileName)) {
-            Long cpuInfoMaxFreq = Long.parseLong(IOUtils.toString(reader).trim());
-            LOGGER.info(String.format("Retrieved value [%s] from file [%s]. This corresponds to a CPU speed of [%s] MHz.", cpuInfoMaxFreq, cpuInfoMaxFreqFileName, cpuInfoMaxFreq / 1000));
-            return cpuInfoMaxFreq / 1000;
-        } catch (IOException | NumberFormatException e) {
-            LOGGER.error(String.format("Unable to retrieve the CPU speed from file [%s]. Using the value [%s] provided by the Libvirt.", cpuInfoMaxFreqFileName, nodeInfo.mhz), e);
-            return nodeInfo.mhz;
+        try {
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:85]: agree with @ravening

**Comment by sureshanaparti** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by davidjumani** [comment]
@blueorangutan help

**Comment by davidjumani** [comment]
@blueorangutan test ubuntu20 ubuntu20

**Comment by davidjumani** [comment]
@blueorangutan test rocky8 kvm-rocky8

**Comment by sureshanaparti** [comment]
@blueorangutan test ubuntu20 kvm-ubuntu20 keepEnv

**Comment by davidjumani** [comment]
Tested on rocky8, suse15 and ubuntu20 as well

**Comment by sureshanaparti** [comment]
Manually verified the CPU speed in CentOS 7.6, Rocky 8.4 and Ubuntu 20.04. Looks good.

Log before changes:
_2022-02-02 16:28:18,088 ERROR [utils.linux.KVMHostInfo] (Agent-Handler-1:null) (logid:e95232e8) Unable to retrieve the CPU speed from file [/sys/devices/system/cpu/cpu0/cpufreq/cpuinfo_max_freq]. Using the value [2100] provided by the Libvirt._

Log after changes:
_2022-02-02 17:00:32,614 INFO  [utils.linux.KVMHostInfo] (Agent-Handler-1:null) (logid:e3ee9c00) Fetching CPU speed from command "lscpu".
2022-02-02 17:00:32,639 INFO  [utils.linux.KVMHostInfo] (Agent-Handler-1:null) (logid:e3ee9c00) Command [lscpu | grep -i 'Model name' | head -n 1 | egrep -o '[[:digit:]].[[:digit:]]+GHz' | sed 's/GHz//g'] resulted in the value [2100] for CPU speed._

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 7]
```diff
@@ -80,16 +82,39 @@ public long getOverCommitMemory() {
...
+        LOGGER.info("Fetching CPU speed from command \"lscpu\".");
+        String command = "lscpu | grep -i 'Model name' | head -n 1 | egrep -o '[[:digit:]].[[:digit:]]+GHz' | sed 's/GHz//g'";
+        String result = Script.runSimpleBashScript(command);
+        long speed = (long) (Float.parseFloat(result) * 1000);
+        LOGGER.info(String.format("Command [%s] resulted in the value [%s] for CPU speed.", command, speed));
+        return speed;
+    }
+
+    private static long getCpuSpeedFromFile() throws IOException {
+        Reader reader = new FileReader(cpuInfoMaxFreqFileName);
```
Comment (weizhouapache) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:112]: add a log info here ?
```
        LOGGER.info("Fetching CPU speed from file \"" + cpuInfoMaxFreqFileName + "\".");
```

File: plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java [hunk 8]
```diff
@@ -80,16 +82,39 @@ public long getOverCommitMemory() {
...
+    private static long getCpuSpeedInternal() throws IOException {
+        try {
+            return getCpuSpeedFromCommandLscpu();
+        } catch (NullPointerException | NumberFormatException e) {
+            LOGGER.error(String.format("Unable to retrieve the CPU speed from lscpu. Using the value in [%s].", cpuInfoMaxFreqFileName), e);
+            return getCpuSpeedFromFile();
+        }
+    }
+
+    private static long getCpuSpeedFromCommandLscpu() {
```
Comment (weizhouapache) [inline plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/utils/linux/KVMHostInfo.java:102]: suggestion:
add try...catch in getCpuSpeedFromCommandLscpu and getCpuSpeedFromFile
if exception is caught, then return the a default value like 0 or -1.

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by davidjumani** [comment]
@blueorangutan package

**Comment by sureshanaparti** [comment]
@blueorangutan test

**Comment by rohityadavcloud** [comment]
ping @davidjumani is this ready, or already reviewed/tested manually @sureshanaparti @nvazquez @DaanHoogland ?

**Comment by DaanHoogland** [comment]
2 lgtm and tested; merging

**Comment by sureshanaparti** [comment]
> ping @davidjumani is this ready, or already reviewed/tested manually @sureshanaparti @nvazquez @DaanHoogland ?

yes, verified the latest changes on CentOS 7.6 and Rocky 8.4.

**Comment by davidjumani** [comment]
@blueorangutan help

**Comment by loth** [comment]
FYI libvirt mentions not to use this API call to get mhz, see https://libvirt.org/html/libvirt-libvirt-host.html#virNodeGetInfo @weizhouapache

**Comment by weizhouapache** [comment]
@loth hmm, thanks Tyler

cc @davidjumani @nvazquez

**Pull request merged**